Rework nominations

ks129 opened the pull request:
**Note! This HAVE to be merged EXACTLY the same time as bot PR**

# Changes
Instead of one nominations table, there are now 2 tables: Nominations and nomination entries. The nominations table got one new field `reviewed` (boolean). Here are schemas of these tables:
```
Table: nomination

active (boolean)
user (User)
inserted_at (datetime)
end_reason (string)
ended_at (datetime)
reviewed (boolean)
```
```
Table: nomination entries

nomination (Nomination)
actor (User)
reason (string)
inserted_at (datetime)
```

Now is also allowed multiple nominations per-user and they are created as nomination entries. One user can create only 1 entry for each nomination.

Also updated tests to match with all these changes.

# What to test?
- Is the migration script properly moving data from old schema to new.
- Is GET routes working as expected?
- Is the POST route creating a new nomination entry if nomination already exists?
- Is the PATCH route `reviewed` field setting works?
- Is the PATCH patching behaviors correct?
- Is tests covering everything (there is one place where I had to ignore it)?

Closes #445

ks129 committed:
Split nomination model to 2 tables and create migrations

ks129 committed:
Change nominations admin interface and add nomination entries interface

ks129 committed:
Change nominations serializer and add nomination entry serializer

ks129 committed:
Update nominations viewset GET and POST to make this working with 2-table system

ks129 committed:
Add reviewed field to nomination serializer

ks129 committed:
Migrate PATCH request for 2-table nominations system

ks129 committed:
Disable creating multiple nomination entries of one nomination for one actor

ks129 committed:
Add rollback to nominations table split migration

ks129 committed:
Wrap validation errors to []

ks129 committed:
Import NominationEntry to models __init__.py

ks129 committed:
Change NominationEntry import location in serializers

ks129 committed:
Small improvements in nomination viewset

ks129 committed:
Update nomination-related tests to cover recent table-splitting changes

SebastiaanZ commented on pydis_site/apps/api/models/bot/nomination.py:
```diff
@@ -52,3 +45,33 @@ class Meta:
+    reason = models.TextField(
+        help_text="Why the actor nominated this user.",
+        null=True,
+        blank=True
```
Do we want two different options to represent an empty reason, `""` and `null`? Could you explain why? The Django docs recommend you to not use `null=True` for character-based fields, as they already have a "logical" null value being an empty string.

SebastiaanZ commented on pydis_site/apps/api/models/bot/nomination.py:
```diff
@@ -52,3 +45,33 @@ class Meta:
+        help_text="The creation date of this nomination entry."
+    )
+
+    class Meta:
+        """Meta options for NominationEntry model."""
+
+        verbose_name_plural = "nomination entries"
```
It could be useful to specify a default [`ordering`](https://docs.djangoproject.com/en/3.1/ref/models/options/#ordering) here that uses the `inserted_at` date, either ascending or descending. This makes it so that nominations can be displayed in chronological order (or reverse chronological order) without having to specify anything anywhere else.

SebastiaanZ commented on pydis_site/apps/api/serializers.py:
```diff
@@ -338,6 +339,21 @@ def create(self, validated_data: dict) -> User:
+class NominationEntrySerializer(ModelSerializer):
+    """A class providing (de-)serialization of `NominationEntry` instances."""
+
+    nomination = PrimaryKeyRelatedField(
+        queryset=Nomination.objects.all(),
+        write_only=True
+    )
```
Could you explain what this does? I'd expected to see a forward `PrimaryKeyRelatedField` serializer relationship in which a `Nomination` has a list with all the `NominationEntries` that belong to that specific nomination, like the `Album` model in the example in the docs having a `many=True` `PrimaryKeyRelatedField` relationship to the individual tracks on the album.

Don't we have something very similar? A single Nomination with several NominationEntries like an Album has many individual Tracks?

Ideally, I'd like to see reasoning for the choice you've made in your commit message, but that may now be hard to add after the fact, as it would rewrite history.

ks129 replied:
I don't want to include nomination ID in the read result, but if I don't define queryset, I get `AssertionError: Relational field must provide a queryset argument, override get_queryset, or set read_only=True.`.

SebastiaanZ commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -148,10 +160,44 @@ class NominationViewSet(CreateModelMixin, RetrieveModelMixin, ListModelMixin, Ge
+                NominationEntry.objects.filter(nomination_id=nomination["id"]),
+                many=True
+            ).data
+            data[i]["entries"] = entries
```
Wouldn't this be solved by adding a `serializers.PrimaryKeyRelatedField` to the `NominationSerializer`? That saves us from doing a manual `for`-loop here. In fact, it probably means we wouldn't have to overwrite the list method at all, as we don't have need for custom functionality that way.

SebastiaanZ commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -148,10 +160,44 @@ class NominationViewSet(CreateModelMixin, RetrieveModelMixin, ListModelMixin, Ge
+            many=True
+        ).data
+
+        return Response(data)
```
My guess is that we don't need this if we make sure to serialize a `Nomination` with its related `NominationEntry` information in the serializer part of the framework. Saves us from overwriting a method.

SebastiaanZ reviewed (changes requested):
I ran out of time for now, but I'm a few commits in. I've just gone over the code as of now and I wanted to give my feedback early, rather than late, as I'm not sure when I'll have time again. 

Things I haven't reviewed yet:
- Grammar
- Functional test

As a general comment: None of the commit messages I've read have a meaningful commit message body and some of the titles are a bit too long. Ideally, a commit message should explain the what and why of the change you're committing to the history instead of having just a single summary line.

Some of these commits are fairly sizeable and you've made a number of choices within those commits; try to add the reasoning behind those choices into the commit message body so that reasoning becomes part of the git history of the repository.

These are few links @lemonsaurus typically shares when talking about commit messages:

https://chris.beams.io/posts/git-commit/
https://thoughtbot.com/blog/5-useful-tips-for-a-better-commit-message
http://ablogaboutcode.com/2011/03/23/proper-git-commit-messages-and-an-elegant-git-history
https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

A more extreme example: https://dhwthompson.com/2019/my-favourite-git-commit

This page is also really nice: https://mtlynch.io/code-review-love/

ks129 committed:
Set related_name option of nomination field of NominationEntry

In order to use entries in serializer without manually
setting entries key we have to use related_name option to
automatically fetch all related entries.

ks129 committed:
Set default ordering of NominationEntry to inserted_at decreasing

Set it here so we don't have to set it every place where we
fetch entries.

ks129 committed:
Merge nomination migrations

This doesn't make sense to have 3 small migrations for one PR,
so I merged 2 existing migrations and 1 new, ordering and related_name adding migrations to one.

ks129 committed:
Add entries field to Nomination serializer

After setting related_name in NominationEntry model nomination field,
we can just provide serializer and DRF automatically fetch
all related entries.

ks129 committed:
Simplify nominations viewset

After moving entries to nomination serializer we can get rid
from GET request handlers and let DRF handle this.
Also PATCH and POST handlers got some simplification by
removing manual entries setting.

ks129 committed:
Remove GET nomination test for entries

Entries isn't handled manually anymore so these tests have no point.

ks129 committed:
Make default value of nomination entry reason to empty string

For string fields NULL as default is not suggested, so use empty string instead.

ks129 committed:
Add comment about manually defining nomination field of NominationEntry serializer

lemonsaurus commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -75,8 +86,9 @@ class NominationViewSet(CreateModelMixin, RetrieveModelMixin, ListModelMixin, Ge
     and `actor` need to know by the site. Providing other valid fields
-    is not allowed and invalid fields are ignored. A `user` is only
-    allowed one active nomination at a time.
+    is not allowed and invalid fields are ignored. If `user` already have
```
```suggestion
    is not allowed and invalid fields are ignored. If `user` already has an
```

lemonsaurus commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -75,8 +86,9 @@ class NominationViewSet(CreateModelMixin, RetrieveModelMixin, ListModelMixin, Ge
-    is not allowed and invalid fields are ignored. A `user` is only
-    allowed one active nomination at a time.
+    is not allowed and invalid fields are ignored. If `user` already have
+    active nomination, new nomination entry will be created assigned to
```
```suggestion
    active nomination, a new nomination entry will be created and assigned as the
```

jb3 commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -221,6 +281,34 @@ def partial_update(self, request: HttpRequest, *args, **kwargs) -> Response:
+
+            if not entry_filter.exists():
+                raise ValidationError(
+                    {'actor': ["Actor don't exist or have not nominated user."]}
```
Single quotes.
```suggestion
                    {'actor': ['Actor don't exist or have not nominated user.']}
```

lemonsaurus commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -102,16 +113,18 @@ class NominationViewSet(CreateModelMixin, RetrieveModelMixin, ListModelMixin, Ge
     that providing other valid fields is not allowed and invalid fields are ignored.
 
-    ### 1. Updating the `reason` of `active` nomination
+    ### 1. Updating the `reason` of `active` nomination. Actor field is required.
```
```suggestion
    ### 1. Updating the `reason` of `active` nomination. The `actor` field is required.
```

jb3 commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -163,19 +187,50 @@ def create(self, request: HttpRequest, *args, **kwargs) -> Response:
+                actor__id=entry_serializer.validated_data["actor"].id
+        ).exists():
+            raise ValidationError(
+                {'actor': ['This actor have already created nomination entry for this nomination.']}
```
Grammar
```suggestion
                {'actor': ['This actor has already endorsed this nomination.']}
```

SebastiaanZ replied:
If this is going to be changed, it's needs to be changed in the test as well

lemonsaurus commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -163,19 +187,50 @@ def create(self, request: HttpRequest, *args, **kwargs) -> Response:
-            data=ChainMap(
-                request.data,
-                {"active": True}
+            # Serializer truncate unnecessary data away
```
```suggestion
            # The serializer will truncate and get rid of excessive data
```

jb3 commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -213,6 +268,11 @@ def partial_update(self, request: HttpRequest, *args, **kwargs) -> Response:
 
+            if 'reviewed' in request.data:
+                raise ValidationError(
+                    {'reviewed': ['This field cannot be set same time than ending nomination.']}
```
I don't understand what this means, but I think this is an improvement. (It may be the opposite).
```suggestion
                    {'reviewed': ['This field cannot be set while you are ending a nomination.']}
```

SebastiaanZ replied:
Make sure to change the test as well.

jb3 commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -221,6 +281,34 @@ def partial_update(self, request: HttpRequest, *args, **kwargs) -> Response:
+        if 'reason' in request.data:
+            if 'actor' not in request.data:
+                raise ValidationError(
+                    {'actor': ['This field is required when editing reason.']}
```
```suggestion
                    {'actor': ['This field is required when editing the reason.']}
```

jb3 commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -221,6 +281,34 @@ def partial_update(self, request: HttpRequest, *args, **kwargs) -> Response:
+            # 4. We're setting nomination reviewed
+            if not instance.active:
+                raise ValidationError(
+                    {'reviewed': ['This field cannot be set if nomination is inactive.']}
```
```suggestion
                    {'reviewed': ['This field cannot be set if the nomination is inactive.']}
```

SebastiaanZ replied:
make sure to adjust the test as well

lemonsaurus reviewed (approved):
I'm mostly okay with this PR. There are grammar issues, but it looks like it does what it needs to do.

jb3 commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -221,6 +281,34 @@ def partial_update(self, request: HttpRequest, *args, **kwargs) -> Response:
 
+        # This is actually covered, but for some reason coverage don't think so.
+        elif 'reviewed' in request.data:  # pragma: no cover
+            # 4. We're setting nomination reviewed
```
```suggestion
            # 4. We are altering the reviewed state of the nomination.
```

jb3 reviewed (approved):
Mainly grammar + code style fixes, but otherwise looks good.

ks129 committed:
Fix grammar of nomination endpoints documentation

Co-authored-by: Leon Sandøy <leon.haland@gmail.com>
Co-authored-by: Joe Banks <joseph@josephbanks.me>

ks129 committed:
Replace double quotes with single quotes

Co-authored-by: Joe Banks <joseph@josephbanks.me>

ks129 committed:
Use double quotes instead apostrophe because string contain "don't"

ks129 committed:
Fix grammar of error messages and change tests to match with changes

mbaruh commented on pydis_site/apps/api/models/bot/nomination.py:
```diff
@@ -42,6 +31,10 @@ class Nomination(ModelReprMixin, models.Model):
     )
+    reviewed = models.BooleanField(
+        default=False,
+        help_text="Whether voting message have been made."
```
Another small correction
```suggestion
        help_text="Whether a review was made."
```

mbaruh commented on pydis_site/apps/api/models/bot/nomination.py:
```diff
@@ -52,3 +45,38 @@ class Meta:
+
+
+class NominationEntry(ModelReprMixin, models.Model):
+    """A nomination entry created by single staff."""
```
```suggestion
    """A nomination entry created by a single staff member."""
```

mbaruh commented on pydis_site/apps/api/models/bot/nomination.py:
```diff
@@ -52,3 +45,38 @@ class Meta:
+    nomination = models.ForeignKey(
+        Nomination,
+        on_delete=models.CASCADE,
+        help_text="Nomination to what this entry belongs.",
```
```suggestion
        help_text="The nomination this entry belongs to.",
```

mbaruh commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -75,8 +86,9 @@ class NominationViewSet(CreateModelMixin, RetrieveModelMixin, ListModelMixin, Ge
-    is not allowed and invalid fields are ignored. A `user` is only
-    allowed one active nomination at a time.
+    is not allowed and invalid fields are ignored. If `user` already has an
+    active nomination, a new nomination entry will be created and assigned as the
```
```suggestion
    active nomination, a new nomination entry will be created and assigned to the
```

mbaruh commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -163,19 +187,50 @@ def create(self, request: HttpRequest, *args, **kwargs) -> Response:
-        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+        entry_serializer.is_valid(raise_exception=True)
+
+        # Don't allow user creating many nomination entries for one nomination
```
```suggestion
        # Don't allow a user to create many nomination entries in a single nomination
```

mbaruh commented on pydis_site/apps/api/viewsets/bot/nomination.py:
```diff
@@ -221,6 +281,34 @@ def partial_update(self, request: HttpRequest, *args, **kwargs) -> Response:
+
+            if not entry_filter.exists():
+                raise ValidationError(
+                    {'actor': ["Actor don't exist or have not nominated user."]}
```
```suggestion
                    {'actor': ["The actor doesn't exist or has not nominated the user."]}
```

mbaruh reviewed (changes requested):
A few more

ks129 committed:
Fix grammar of nomination models

ks129 committed:
Change as -> to in nomination viewset docs

ks129 committed:
Fix grammar of nomination viewset command about single entry for user

ks129 committed:
Fix grammar of unknown actor error and change tests

mbaruh committed:
Merge branch 'master' into ks123/nomination/split
